src/libos/src/net/syscalls.rs

```diff
@@ -200,6 +205,76 @@ pub async fn do_getsockname(
     Ok(0)
 }
 
+pub async fn do_setsockopt(
+    fd: c_int,
+    level: c_int,
+    optname: c_int,
+    optval: *const c_void,
+    optlen: libc::socklen_t,
+) -> Result<isize> {
+    debug!(
+        "setsockopt: fd: {}, level: {}, optname: {}, optval: {:?}, optlen: {:?}",
+        fd, level, optname, optval, optlen
+    );
+
+    let file_ref = current!().file(fd as FileDesc)?;
+    let socket_file = file_ref
+        .as_socket_file()
+        .ok_or_else(|| errno!(EINVAL, "not a socket"))?;
+
+    from_user::check_array(optval, optlen as _)?;
+
+    let mut cmd = unsafe { new_setsockopt_cmd(level, optname, optval, optlen)? };
+    socket_file.ioctl(cmd.as_mut())?;
+    Ok(0)
+}
+
+pub async fn do_getsockopt(
+    fd: c_int,
+    level: c_int,
+    optname: c_int,
+    optval: *mut c_void,
+    optlen: *mut libc::socklen_t,
+) -> Result<isize> {
+    debug!(
+        "getsockopt: fd: {}, level: {}, optname: {}, optval: {:?}, optlen: {:?}",
+        fd, level, optname, optval, optlen
+    );
+
+    let file_ref = current!().file(fd as FileDesc)?;
+    let socket_file = file_ref
+        .as_socket_file()
+        .ok_or_else(|| errno!(EINVAL, "not a socket"))?;
+
+    from_user::check_mut_ptr(optlen)?;
+    from_user::check_mut_array(optval, unsafe { *optlen } as _)?;
```

Comment:
In my opinion, the best practice when handling user raw pointers is to check pointers and then convert them to their Rust counterparts.

```rust
let optlen_mut = {
    from_user::check_mut_ptr(optlen)?;
    &mut *optlen
};
let optlen = *optlen_mut;
let optval_mut = {
    from_user::check_mut_array(optval, optlen as usize)?;
    slice::from_raw_parts_mut(optval as *mut u8, optlen)
};
```

After this point, you can write safe Rust code on `optlen_mut`, `optlen`, and `optval_mut`. This way, the function `put_to_user_unchecked`, `new_getsockopt_cmd`, `new_setsockopt_cmd` won't be unsafe anymore.

I see that you added the safety sections for these unsafe functions. But no explanations can be better than---let's say---no need to explain at all (when unsafe functions do not exist at all).

**Do not write comments to explain bad code. Write good code instead.**